web-console/src/components/segment-timeline.tsx

```diff
@@ -0,0 +1,393 @@
+          let y = 0;
+          if (d[datasource] !== undefined) {
+            y = d[datasource];
+          }
+          return {
+            x: d.day,
+            y,
+            datasource,
+            color: this.colors[i]
+          };
+        });
+        if (!dataResult.every((d: any) => d.y === 0)) singleDatasourceData[dataType][datasource] = dataResult;
+      });
+    });
+    return singleDatasourceData;
+  }
+
+  private calculateScales(): BarChartScales | null {
+    const { chartWidth, chartHeight} = this.props;
+    const {data, activeDataType, activeDatasource, singleDatasourceData } = this.state;
+    if (Object.keys(data).length === 0) return null;
+    const activeData = data[activeDataType];
+    const xDomain: Date[] = [new Date(activeData[0].day), new Date(activeData[activeData.length - 1].day)];
+    let yDomain: number[] = [0, activeData.reduce((max: any, d: any) => max.total > d.total ? max : d).total];
+    if (activeDatasource !== null && singleDatasourceData[activeDataType][activeDatasource] !== undefined) {
+      yDomain = [0, singleDatasourceData[activeDataType][activeDatasource].reduce((max: any, d: any) => max.y > d.y ? max : d).y];
+    }
+    const xScale: AxisScale<Date> = d3.scaleTime().domain(xDomain).range([0, chartWidth - this.chartMargin.left - this.chartMargin.right]);
+    const yScale: AxisScale<number> = d3.scaleLinear().rangeRound([chartHeight - this.chartMargin.top - this.chartMargin.bottom, 0]).domain(yDomain);
+    return {
+      xScale,
+      yScale
+    };
+  }
+
+  onTimeSpanChange = (e: any) => {
+    this.setState({
+      timeSpan: e,
+      loading: true
+    });
+    this.dataQueryManager.rerunLastQuery();
+  }
+
+  formatTick = (n: number) => {
+    const { activeDataType } = this.state;
+    if (activeDataType === 'countData') {
+      return n.toString();
+    } else {
+      return formatBytes(n);
+    }
+  }
+
+  renderStackedBarChart() {
+    const { chartWidth, chartHeight} = this.props;
+    const { loading, dataToRender, activeDataType, error, xScale, yScale, data, activeDatasource } = this.state;
+    if (loading) {
+      return <div>
+        <Loader
+          loading={loading}
+        />
+      </div>;
+    }
+
+    if (error) {
+      return <div>
+        <span className={'no-data-text'}>Error when loading data: {error}</span>
+      </div>;
+    }
+
+    if (xScale === null || yScale === null) {
+      return <div>
+        <span className={'no-data-text'}>Error when calculating scales</span>
+      </div>;
+    }
+
+    if (activeDatasource !== null && data[activeDataType].every((d: any) => d[activeDatasource] === undefined)) {
+      return <div>
+        <span className={'no-data-text'}>No data available for <i>{activeDatasource}</i></span>
+      </div>;
+    }
+
+    const barCounts = data[activeDataType].length;
+    const barWidth = (chartWidth - this.chartMargin.left - this.chartMargin.right) / barCounts;
```

Comment:
I think `barWidth` should probably just be based on the number of days over the x domain, since that appears to be how they are being bucketed. 

I made this change (along with the other suggestion about changing x domain scaling) when testing to see what it might look like:
```
    const dEnd = new Date(new Date().getTime());
    const dStart = new Date(dEnd.getTime());
    dStart.setMonth(dStart.getMonth() - timeSpan);
    const millisecondsPerDay = 24 * 60 * 60 * 1000;
    const barCounts = (dEnd.getTime() - dStart.getTime()) / millisecondsPerDay;
```
<img width="1674" alt="Screen Shot 2019-05-18 at 6 59 52 PM" src="https://user-images.githubusercontent.com/1577461/57976853-4bcd0880-799f-11e9-8313-97e5f0778ae5.png">

<img width="1679" alt="Screen Shot 2019-05-18 at 7 00 08 PM" src="https://user-images.githubusercontent.com/1577461/57976855-512a5300-799f-11e9-991e-9d48e8bce174.png">

If you make these modifications, I would suggest probably calculating the `start` and `end` dates when the timeSpan changes and then re-use for both calculating the 'xDomain' and the 'barCount'.